File: backend/src/main/java/net/alex9849/cocktailmaker/payload/dto/recipe/ManualIngredientDto.java
package net.alex9849.cocktailmaker.payload.dto.recipe;

import lombok.*;
import net.alex9849.cocktailmaker.model.recipe.AutomatedIngredient;
import net.alex9849.cocktailmaker.model.recipe.Ingredient;

@NoArgsConstructor(access = AccessLevel.PRIVATE)
public class ManualIngredientDto {

    @NoArgsConstructor(access = AccessLevel.PRIVATE)
    public static class Request {
        @Getter @Setter @EqualsAndHashCode
        public static class Create extends AddableIngredientDto.Request.Create {
            Ingredient.Unit unit;

            @Override
            public String getType() {
                return "manual";
            }
        }
    }

    @NoArgsConstructor(access = AccessLevel.PRIVATE)
    public static class Response {
        @Getter @Setter @EqualsAndHashCode
        public static class Detailed extends AddableIngredientDto.Response.Detailed {
            Ingredient.Unit unit;

            protected Detailed(AutomatedIngredient ingredient) {
                super(ingredient);
            }

            @Override
            public String getType() {
                return "manual";
            }

        }

        @Getter @Setter @EqualsAndHashCode
        public static class Reduced extends AddableIngredientDto.Response.Reduced {

            protected Reduced(AutomatedIngredient ingredient) {
                super(ingredient);
            }

            @Override
            public String getType() {
                return "manual";
            }
        }
    }
}

File: frontend/src/models/dto/recipe/ingredient/ManualIngredientDto.ts
declare module 'ManualIngredientDto/Request' {
  import AddableIngredientRequest from 'AddableIngredientDto/Request'

  export interface Create extends AddableIngredientRequest.Create {
    type: 'manual';
    unit: string;
  }
}

declare module 'ManualIngredientDto/Response' {
  import AddableIngredientRequest from 'AddableIngredientDto/Response'

  export interface Detailed extends AddableIngredientRequest.Detailed {
    type: 'manual';
  }

  export interface Reduced extends AddableIngredientRequest.Reduced {
    type: 'manual';
  }
}
